# Test the analysis of CKKW-L of DP8

## 1. Import Packages

In [1]:
# The Python Standard Library
import os
import sys
import time
import datetime
import glob
import multiprocessing as mp

# The Third-Party Library
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tqdm
import prettytable
import uproot
import pyjet
import importlib

# My Packages
import myhep.particle_information_v2 as mypiv2
import myhep.analytical_function_v2 as myafv2
import myhep.particleinfo_v1 as mypiv1
import myhep.particlefun_v1 as myafv1
import myhep.parse_ckkwl_DP8 as myparDP8

# increase figure showing resolution
# %config InlineBackend.figure_format = 'retina'

## 2. Import .root Files and Load the Data via class

In [2]:
INPUT_FILE = '/youwei_home/SVJ_CKKWL/s-channel_ckkwl-v2/DP8/svj_ckkwl-4.root'

DATA = uproot.open(INPUT_FILE)['Delphes;1']
GP = mypiv2.classGenParticle(DATA)
Jet = mypiv2.classJet(DATA)
Event = mypiv2.classEvent(DATA)

### 2-1. Check the number of event for each branch

In [3]:
if GP.length == Jet.length == Event.length:
    print("There are {} events in the .root file.".format(GP.length))
else:
    print("""\033[7;31m****** There is the problem for the number of event \
    in the .root file. ******\033[0m""")
#     print("\033[7;31m****** There is the problem for the number of event "
#           "in the .root file. ******\033[0m")
    print("\033[7;31m****** Please check your .root file. ******\033[0m")

There are 10000 events in the .root file.


## 3. Analyze the dark quark pair

In [4]:
df_xdxdx_23 = myafv2.parse_xdxdx_v2(GP, status=23)

IndexError: single positional indexer is out-of-bounds

In [5]:
Event.Weight

<JaggedArray [[1.1548845] [0.0] [2.2193444] ... [1.8988278] [1.1548845] [1.7454567]] at 0x7fd5f938ebb0>

In [6]:
weight_0 = []

for i in range(10000):
    if Event.Weight[i] == 0:
        weight_0.append(i)
        
print(weight_0[0:20])
print(len(weight_0))
print(10000 - len(weight_0))

[1, 13, 15, 17, 22, 23, 27, 32, 33, 46, 48, 50, 55, 58, 62, 68, 70, 72, 74, 75]
2678
7322


### 3-1. Test event with weight = 0

In [7]:
status = 23

In [8]:
dfGP = GP.dataframelize(0)
dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
m1 = dfGP_DQ_Status.iloc[0,6]
m1

10.0

In [9]:
dfGP = GP.dataframelize(1)
dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
m1 = dfGP_DQ_Status.iloc[0,6]
m1

IndexError: single positional indexer is out-of-bounds

In [10]:
dfGP = GP.dataframelize(1)
dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
dfGP_DQ_Status

Empty DataFrame
Columns: [PID, Status, M1, M2, D1, D2, Mass, PT, Eta, Phi]
Index: []

In [11]:
dfGP = GP.dataframelize(13)
dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
dfGP_DQ_Status

Empty DataFrame
Columns: [PID, Status, M1, M2, D1, D2, Mass, PT, Eta, Phi]
Index: []

In [12]:
dfGP = GP.dataframelize(14)
dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
dfGP_DQ_Status

PID  Status   M1  M2   D1   D2  Mass          PT       Eta       Phi
703  4900101      23  540  -1  705  706  10.0  118.455040 -2.883845 -1.932256
704 -4900101      23  540  -1  707  707  10.0  118.590218  2.165266  1.157263

In [13]:
dfGP = GP.dataframelize(15)
dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
dfGP_DQ_Status

Empty DataFrame
Columns: [PID, Status, M1, M2, D1, D2, Mass, PT, Eta, Phi]
Index: []

In [14]:
dfGP = GP.dataframelize(1)
dfGP

PID  Status  M1  M2  D1  D2         Mass          PT         Eta  \
0     2212       4  -1  -1   2  -1     0.938270    0.000000  999.900024   
1     2212       4  -1  -1   3  -1     0.938270    0.000000 -999.900024   
2       -1      21   0  -1   4   6     0.000000    0.000000  999.900024   
3        1      21   1  -1   4   6     0.000000    0.000000 -999.900024   
4  5000001      22   2   3   7   8  1501.464966   70.712898    2.837252   
5       21       1   2   3  -1  -1     0.000000   47.310688   -0.507734   
6       21       1   2   3  -1  -1     0.000000   24.271572   -0.768909   
7  4900101       1   4  -1  -1  -1    10.000000  765.395447    0.241501   
8 -4900101       1   4  -1  -1  -1    10.000000  720.125183    0.548116   

        Phi  
0  0.000000  
1  0.000000  
2  0.000000  
3  0.000000  
4  1.374616  
5 -1.655615  
6 -1.985327  
7  2.214620  
8 -0.853787

In [15]:
dfGP = GP.dataframelize(13)
dfGP

PID  Status  M1  M2  D1  D2         Mass          PT         Eta  \
0     2212       4  -1  -1   2  -1     0.938270    0.000000  999.900024   
1     2212       4  -1  -1   3  -1     0.938270    0.000000 -999.900024   
2       -1      21   0  -1   4   5     0.000000    0.000000  999.900024   
3        1      21   1  -1   4   5     0.000000    0.000000 -999.900024   
4  5000001      22   2   3   6   7  1484.444824   56.841114   -3.697461   
5       21       1   2   3  -1  -1     0.000000   56.841114   -2.132070   
6  4900101       1   4  -1  -1  -1    10.000000  388.211212    0.568831   
7 -4900101       1   4  -1  -1  -1    10.000000  352.831329   -2.072141   

        Phi  
0  0.000000  
1  0.000000  
2  0.000000  
3  0.000000  
4  2.466349  
5 -0.675244  
6  1.626104  
7 -1.635767

In [16]:
dfGP = GP.dataframelize(14)
dfGP

PID  Status    M1  M2    D1    D2         Mass        PT  \
0        2212       4    -1  -1   538    -1     0.938270  0.000000   
1        2212       4    -1  -1   539    -1     0.938270  0.000000   
2           2      21    43  43     4    -1     0.000000  0.000000   
3          -2      21    44  -1     4    -1     0.000000  0.000000   
4     5000001      22     2   3    45    45  1489.427246  0.000000   
...       ...     ...   ...  ..   ...   ...          ...       ...   
2032      111       2  2024  -1  2035  2036     0.134980  0.456895   
2033       22       1  2028  -1    -1    -1     0.000000  0.386439   
2034       22       1  2028  -1    -1    -1     0.000000  1.158487   
2035       22       1  2032  -1    -1    -1     0.000000  0.043480   
2036       22       1  2032  -1    -1    -1     0.000000  0.417836   

             Eta       Phi  
0     999.900024  0.000000  
1    -999.900024  0.000000  
2     999.900024  0.000000  
3    -999.900024  0.000000  
4    -999.900024  0.000000  
...          ...       ...  
2032   -1.653047 -1.432048  
2033   -2.048207  2.934754  
2034   -2.244369  2.980519  
2035   -0.845828 -0.998522  
2036   -1.702239 -1.475775  

[2037 rows x 10 columns]

In [17]:
dfGP = GP.dataframelize(15)
dfGP

PID  Status  M1  M2  D1  D2         Mass          PT         Eta  \
0     2212       4  -1  -1   2  -1     0.938270    0.000000  999.900024   
1     2212       4  -1  -1   3  -1     0.938270    0.000000 -999.900024   
2       -5      21   0  -1   4   5     0.000000    0.000000  999.900024   
3       21      21   1  -1   4   5     0.000000    0.000000 -999.900024   
4  5000001      22   2   3   6   7  1432.370361   23.177889   -1.526180   
5       -5       1   2   3  -1  -1     4.800000   23.177889   -2.541455   
6  4900101       1   4  -1  -1  -1    10.000000  715.149048    0.026939   
7 -4900101       1   4  -1  -1  -1    10.000000  714.482971   -0.097907   

        Phi  
0  0.000000  
1  0.000000  
2  0.000000  
3  0.000000  
4 -2.679239  
5  0.462354  
6 -1.153385  
7  2.020620

#### Concusion: Because there are 2678 events without Status=23 4900101 and no hadronization.

## 4. Test New Analysis for DP8

In [18]:
event_weight_not0 = []

for i in range(GP.length):
    if Event.Weight[i] != 0:
        event_weight_not0.append(i)
        
print(event_weight_not0[0:20])
print(len(event_weight_not0))

[0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 16, 18, 19, 20, 21, 24, 25]
7322


In [19]:
# 1. Invariant Mass and Transverse Mass
def M(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2):
    px1, py1, pz1 = pt1*np.cos(phi1), pt1*np.sin(phi1), np.sqrt(m1**2+pt1**2)*np.sinh(eta1)
    e1 = np.sqrt(m1**2 + px1**2 + py1**2 + pz1**2)
    px2, py2, pz2 = pt2*np.cos(phi2), pt2*np.sin(phi2), np.sqrt(m2**2+pt2**2)*np.sinh(eta2)
    e2 = np.sqrt(m2**2 + px2**2 + py2**2 + pz2**2)
    return np.sqrt((e1+e2)**2 - (px1+px2)**2 - (py1+py2)**2 - (pz1+pz2)**2)

def MT(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2):
    px1, py1, pz1 = pt1*np.cos(phi1), pt1*np.sin(phi1), np.sqrt(m1**2+pt1**2)*np.sinh(eta1)
    e1 = np.sqrt(m1**2 + px1**2 + py1**2 + pz1**2)
    px2, py2, pz2 = pt2*np.cos(phi2), pt2*np.sin(phi2), np.sqrt(m2**2+pt2**2)*np.sinh(eta2)
    e2 = np.sqrt(m2**2 + px2**2 + py2**2 + pz2**2)
    ET1, ET2 = np.sqrt(m1**2 + pt1**2), np.sqrt(m2**2 + pt2**2)
    return np.sqrt((ET1+ET2)**2 - (px1+px2)**2 - (py1+py2)**2)

# 9. Transverse Mass mT is invariant under Lorentz boost along the z direction.
def mT12(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2):
    px1, py1, pz1 = pt1*np.cos(phi1), pt1*np.sin(phi1), np.sqrt(m1**2+pt1**2)*np.sinh(eta1)
    e1 = np.sqrt(m1**2 + px1**2 + py1**2 + pz1**2)
    px2, py2, pz2 = pt2*np.cos(phi2), pt2*np.sin(phi2), np.sqrt(m2**2+pt2**2)*np.sinh(eta2)
    e2 = np.sqrt(m2**2 + px2**2 + py2**2 + pz2**2)
    return np.sqrt((e1+e2)**2 - (pz1+pz2)**2)

# 14. Transverse Energy ET
def ET(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2):
    px1, py1, pz1 = pt1*np.cos(phi1), pt1*np.sin(phi1), np.sqrt(m1**2+pt1**2)*np.sinh(eta1)
    e1 = np.sqrt(m1**2 + px1**2 + py1**2 + pz1**2)
    px2, py2, pz2 = pt2*np.cos(phi2), pt2*np.sin(phi2), np.sqrt(m2**2+pt2**2)*np.sinh(eta2)
    e2 = np.sqrt(m2**2 + px2**2 + py2**2 + pz2**2)
    m12 = np.sqrt((e1+e2)**2 - (px1+px2)**2 - (py1+py2)**2 - (pz1+pz2)**2)
    return np.sqrt(m12**2 + (px1+px2)**2 + (py1+py2)**2)

In [20]:
# 15. Parse the Dark Quark Pair xd and xdx
def parse_xdxdx_v2_ckkwl_DP8(GP, status=23):
    """
    GP=GenParticle; i=i-th event; df=dataframe; Tem=temporary; acc=accumulate
    """
    _list_M, _list_MT, _list_mT, _list_ET = [], [], [], []
    _list_Dphi, _list_Deta = [], []
    _error = []
    acc = 0
    for i in event_weight_not0:
        dfGP = GP.dataframelize(i)
        dfGP_DQ_Status = dfGP[(abs(dfGP['PID']) == 4900101) & (dfGP['Status'] == status)]
        m1 = dfGP_DQ_Status.iloc[0,6]
        pt1 = dfGP_DQ_Status.iloc[0,7]
        eta1 = dfGP_DQ_Status.iloc[0,8]
        phi1 = dfGP_DQ_Status.iloc[0,9]
        m2 = dfGP_DQ_Status.iloc[1,6]
        pt2 = dfGP_DQ_Status.iloc[1,7]
        eta2 = dfGP_DQ_Status.iloc[1,8]
        phi2 = dfGP_DQ_Status.iloc[1,9]
        
        _list_M.append(M(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2))
        _list_MT.append(MT(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2))
        _list_mT.append(mT12(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2))
        _list_ET.append(ET(m1, pt1, eta1, phi1, m2, pt2, eta2, phi2))
        Dphi, Deta = abs(phi1 - phi2), abs(eta1 - eta2)
        if Dphi > np.pi:
            _list_Dphi.append(2*np.pi - Dphi)
        else:
            _list_Dphi.append(Dphi)
        _list_Deta.append(Deta)
        
        
        if dfGP_DQ_Status.shape[0] != 2:
            acc += 1
            _error.append(i)
            
    M_xdxdx, MT_xdxdx = np.array(_list_M), np.array(_list_MT)
    mT_xdxdx, ET_xdxdx = np.array(_list_mT), np.array(_list_ET)
    Dphi_xdxdx, Deta_xdxdx = np.array(_list_Dphi), np.array(_list_Deta)
    data_xdxdx = {"M_xdxdx": M_xdxdx, "MT_xdxdx": MT_xdxdx,
                  "mT_xdxdx": mT_xdxdx, "ET_xdxdx": ET_xdxdx,
                  "Dphi_xdxdx": Dphi_xdxdx, "Deta_xdxdx": Deta_xdxdx}
    df_xdxdx = pd.DataFrame(data_xdxdx)
    if acc == 0:
        print("All events only include 2 particles.")
    else:
        print("There are {} events with over 2 particles.".format(acc))
        print(_error)
    return df_xdxdx

In [21]:
df_xdxdx_23_test = parse_xdxdx_v2_ckkwl_DP8(GP, status=23)

All events only include 2 particles.


In [22]:
df_xdxdx_23_test

M_xdxdx     MT_xdxdx     mT_xdxdx     ET_xdxdx  Dphi_xdxdx  \
0     1507.636644  1507.462321  1512.711969  1512.711969    3.014711   
1     1468.239444  1081.752578  1468.647231  1468.647231    3.096193   
2     1504.344619  1498.178967  1504.621514  1504.621514    3.123330   
3     1506.168102  1089.146554  1506.204504  1506.204504    3.131495   
4     1533.388259  1201.108746  1539.448959  1539.448959    3.127897   
...           ...          ...          ...          ...         ...   
7317  1498.032041   475.409804  1552.484533  1552.484533    2.932542   
7318  1529.893599   808.049474  1670.261296  1670.261296    1.801678   
7319  1563.256152  1227.471671  1566.392100  1566.392100    2.998684   
7320  1530.403208   484.306907  1540.942646  1540.942646    2.429980   
7321  1463.710062  1147.137303  1471.155837  1471.155837    2.920415   

      Deta_xdxdx  
0       0.030355  
1       1.643608  
2       0.181378  
3       1.698638  
4       1.455313  
...          ...  
7317    3.617849  
7318    2.108055  
7319    1.444381  
7320    3.513257  
7321    1.446043  

[7322 rows x 6 columns]

#### Conclusion: Therefore, I need to additionally append the valid event list.

## 5. Test the New Analysis Script = analysis-1_ckkwl_DP8.py

### 5-3. Analyze the Dark Quark Pair

In [23]:
df_xdxdx_23 = myparDP8.parse_xdxdx_v2_ckkwl_DP8(GP, event_weight_not0, status=23)
df_xdxdx_71 = myparDP8.parse_xdxdx_v2_ckkwl_DP8(GP, event_weight_not0, status=71)

All events only include 2 particles.
All events only include 2 particles.


In [24]:
df_xdxdx_23

M_xdxdx     MT_xdxdx     mT_xdxdx     ET_xdxdx  Dphi_xdxdx  \
0     1507.636644  1507.462321  1512.711969  1512.711969    3.014711   
1     1468.239444  1081.752578  1468.647231  1468.647231    3.096193   
2     1504.344619  1498.178967  1504.621514  1504.621514    3.123330   
3     1506.168102  1089.146554  1506.204504  1506.204504    3.131495   
4     1533.388259  1201.108746  1539.448959  1539.448959    3.127897   
...           ...          ...          ...          ...         ...   
7317  1498.032041   475.409804  1552.484533  1552.484533    2.932542   
7318  1529.893599   808.049474  1670.261296  1670.261296    1.801678   
7319  1563.256152  1227.471671  1566.392100  1566.392100    2.998684   
7320  1530.403208   484.306907  1540.942646  1540.942646    2.429980   
7321  1463.710062  1147.137303  1471.155837  1471.155837    2.920415   

      Deta_xdxdx  
0       0.030355  
1       1.643608  
2       0.181378  
3       1.698638  
4       1.455313  
...          ...  
7317    3.617849  
7318    2.108055  
7319    1.444381  
7320    3.513257  
7321    1.446043  

[7322 rows x 6 columns]

In [25]:
df_xdxdx_71

M_xdxdx     MT_xdxdx     mT_xdxdx     ET_xdxdx  Dphi_xdxdx  \
0     1028.760900  1028.742243  1063.729147  1063.729147    3.090966   
1      627.788842   371.290808   663.741608   663.741608    2.121389   
2      872.789634   868.344030   889.244642   889.244642    3.090246   
3      335.110276    42.071662   383.155727   383.155727    0.479586   
4     1243.134252   985.810764  1259.966047  1259.966047    2.906756   
...           ...          ...          ...          ...         ...   
7317   769.703241   308.301038   850.113713   850.113713    3.125033   
7318   552.178556    43.091309   849.156910   849.156910    0.119417   
7319  1081.861620   865.829095  1091.241998  1091.241998    3.009926   
7320  1263.659323   404.778161  1271.958083  1271.958083    2.462261   
7321   917.067624   761.178800   922.493599   922.493599    2.947320   

      Deta_xdxdx  
0       0.012050  
1       2.019419  
2       0.202224  
3       3.279907  
4       1.406944  
...          ...  
7317    3.125296  
7318    1.563178  
7319    1.382381  
7320    3.498887  
7321    1.253840  

[7322 rows x 6 columns]

### 5-4. Select Stable Final State Particle and Filter out Dark Matter

In [26]:
list_SFSP, list_SFSP_filterDM = myparDP8.selectStableFinalStateParticle_filterDM(
    GP, event_weight_not0, FILTER=[51, -51, 53, -53])

The PID of dark matter are [51, -51, 53, -53].
There are 7322 events with stable final state.
There are 7322 events with stable final state and without DM.


In [28]:
list_SFSP[0]

array([[ 1.85465679e-01,  7.25326896e-01,  3.27787817e-01, ...,
         8.82298768e-01,  5.92146739e-02,  4.08539772e-02],
       [-7.36003828e+00, -8.19876957e+00, -7.11297607e+00, ...,
        -5.06325424e-01,  7.14360893e-01, -1.19961154e+00],
       [-2.26445532e+00,  2.54555655e+00, -3.13126385e-01, ...,
        -2.74682951e+00,  1.36656988e+00, -3.03906655e+00],
       [ 1.39569998e-01,  1.39569998e-01,  1.39569998e-01, ...,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
       [ 2.11000000e+02, -2.11000000e+02,  2.11000000e+02, ...,
         2.20000000e+01,  2.20000000e+01,  2.20000000e+01]])

In [29]:
list_SFSP_filterDM[0]

array([[ 1.85465679e-01,  7.25326896e-01,  3.27787817e-01, ...,
         8.82298768e-01,  5.92146739e-02,  4.08539772e-02],
       [-7.36003828e+00, -8.19876957e+00, -7.11297607e+00, ...,
        -5.06325424e-01,  7.14360893e-01, -1.19961154e+00],
       [-2.26445532e+00,  2.54555655e+00, -3.13126385e-01, ...,
        -2.74682951e+00,  1.36656988e+00, -3.03906655e+00],
       [ 1.39569998e-01,  1.39569998e-01,  1.39569998e-01, ...,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
       [ 2.11000000e+02, -2.11000000e+02,  2.11000000e+02, ...,
         2.20000000e+01,  2.20000000e+01,  2.20000000e+01]])

### 5-5. Let's Do the Jet Clustering!!

In [31]:
R = 0.4
jetClusteringAlgorithm = -1
pTmin_pyjet = 0

list_PseudoJet = myafv2.jetClustering(list_SFSP, R=R,
                                      p=jetClusteringAlgorithm,
                                      pTmin=pTmin_pyjet)
list_PseudoJet_filterDM = myafv2.jetClustering(list_SFSP_filterDM, R=R,
                                               p=jetClusteringAlgorithm,
                                               pTmin=pTmin_pyjet)

In [40]:
list_PseudoJet[0][0:5]

[PseudoJet(pt=728.984, eta=-0.247, phi=-2.799, mass=97.999),
 PseudoJet(pt=708.831, eta=-0.262, phi=0.463, mass=60.708),
 PseudoJet(pt=80.476, eta=1.887, phi=-1.083, mass=9.673),
 PseudoJet(pt=57.552, eta=-1.910, phi=0.581, mass=10.780),
 PseudoJet(pt=34.003, eta=0.045, phi=3.089, mass=5.587)]

In [39]:
list_PseudoJet_filterDM[0][0:5]

[PseudoJet(pt=570.051, eta=-0.252, phi=-2.796, mass=77.598),
 PseudoJet(pt=518.551, eta=-0.267, phi=0.455, mass=38.772),
 PseudoJet(pt=80.476, eta=1.887, phi=-1.083, mass=9.673),
 PseudoJet(pt=57.552, eta=-1.910, phi=0.581, mass=10.780),
 PseudoJet(pt=34.003, eta=0.045, phi=3.089, mass=5.587)]

### 5-6. Analyze the Jet in the Truth Level

In [41]:
N_jet = 2

list_truth_jet = myafv2.preselectTruthJet(list_PseudoJet, N_JET_MIN=N_jet)
list_truth_jet_filterDM = myafv2.preselectTruthJet(list_PseudoJet_filterDM,
                                                   N_JET_MIN=N_jet)

The order of jets all are no error!!
There are 7322 selected events.
The order of jets all are no error!!
There are 7322 selected events.


In [42]:
list_truth_jet

[array([1444.24949057,  800.25664476, 1448.981974  , ..., 1480.51192189,
        1195.14268779, 1188.20297817]),
 array([1444.20809938,  694.60474261, 1443.61381775, ..., 1183.55370502,
         389.23389739,  988.55260242]),
 array([1447.00879733,  838.41342761, 1450.18729489, ..., 1483.54801333,
        1201.47308886, 1192.93575932]),
 array([   0,    1,    2, ..., 7319, 7320, 7321]),
 array([728.98362925, 426.79599686, 744.13915168, ..., 616.15879902,
        203.50024788, 514.38461711]),
 array([708.83137275, 296.79717042, 685.68641073, ..., 564.95693271,
        200.48208566, 462.13749999]),
 array([3.02046775, 2.5320362 , 3.12925191, ..., 3.0061348 , 2.52210752,
        2.95177238])]

In [43]:
list_truth_jet_filterDM

[array([1092.50535411,  716.95029638,  467.63684472, ..., 1032.06891657,
        1195.14268779, 1141.57432267]),
 array([1092.47446063,  433.60423817,  466.56649554, ...,  826.29032984,
         389.23389739,  949.23105039]),
 array([1095.33719147,  770.9670521 ,  586.11580163, ..., 1067.00654388,
        1201.47308886, 1148.45095087]),
 array([   0,    1,    2, ..., 7319, 7320, 7321]),
 array([570.05066742, 335.59698341, 465.20504585, ..., 564.95693271,
        203.50024788, 514.38461711]),
 array([518.55096828, 175.30181099, 112.16869985, ..., 299.03788128,
        200.48208566, 431.63764265]),
 array([3.03205221, 2.13565298, 3.07771404, ..., 3.01689052, 2.52210752,
        2.94103748])]

### 5-7. Analyze the MET in the Truth Level

In [44]:
list_truth_MET = myafv2.parseMETTruthJet(list_SFSP_filterDM,
                                         list_PseudoJet_filterDM,
                                         N_jet_min=N_jet)

There are 7322 invisible events.
There are 7322 jet events.


In [45]:
list_truth_MET

[array([ 41.40285417,  75.87466047, 419.54760045, ..., 317.15452317,
         38.45383419,  31.26549214]),
 array([1.12231463, 3.04668158, 4.89435522, ..., 3.8682515 , 4.72896339,
        5.58274997]),
 array([1133.22551133,  820.31901803, 1003.46845797, ..., 1380.90647769,
        1229.55313074, 1174.93911152]),
 array([2.3644486 , 2.36448407, 3.10803763, ..., 2.99599344, 1.56580607,
        3.09812707]),
 array([   0,    1,    2, ..., 7319, 7320, 7321])]

### 5-8. Analyze the Jet in the Detector Level

In [46]:
pT1_min = 440
pT2_min = 60
eta12_max = 1.2

list_detector_jet = myparDP8.preselectDetectorJet(Jet,
                                                  event_weight_not0,
                                                  N_JET_MIN=N_jet,
                                                  PT1_MIN=pT1_min,
                                                  PT2_MIN=pT2_min,
                                                  ETA_MAX=eta12_max)

The order of jets all are no error!!
There are 2819 selected events.


In [47]:
list_detector_jet

[array([1301.57285351, 1064.6210029 ,  982.55653551, ..., 1083.73875808,
        1089.9781871 , 1163.77880126]),
 array([1301.43493313, 1059.7557313 ,  975.30195833, ..., 1044.77612102,
         960.36560203,  981.41101771]),
 array([1304.55753221, 1075.68102455,  993.43047459, ..., 1100.18890544,
        1091.19315081, 1168.89974414]),
 array([   0,    3,    6, ..., 9991, 9992, 9999]),
 array([659.62823, 604.8492 , 552.1388 , ..., 609.7374 , 501.84848,
        513.98175], dtype=float32),
 array([631.54236, 451.47757, 416.54233, ..., 426.89542, 450.40314,
        456.24863], dtype=float32),
 array([3.01196987, 3.11823875, 3.02542186, ..., 3.04365396, 3.1377011 ,
        2.9496575 ])]

### 5-9. Define the Event Weight

In [49]:
nevents = 10000

event_weight = np.array(Event.Weight)/nevents

In [50]:
event_weight

array([array([0.00011549], dtype=float32), array([0.], dtype=float32),
       array([0.00022193], dtype=float32), ...,
       array([0.00018988], dtype=float32),
       array([0.00011549], dtype=float32),
       array([0.00017455], dtype=float32)], dtype=object)

In [54]:
np.sum(event_weight)[0]

1.1259748